comp/netflow/payload/payload.go

```diff
@@ -39,25 +41,73 @@ type ObservationPoint struct {
-	Packets        uint64           `json:"packets"`
-	EtherType      string           `json:"ether_type,omitempty"`
-	IPProtocol     string           `json:"ip_protocol"`
-	Device         Device           `json:"device"`
-	Exporter       Exporter         `json:"exporter"`
-	Source         Endpoint         `json:"source"`
-	Destination    Endpoint         `json:"destination"`
-	Ingress        ObservationPoint `json:"ingress"`
-	Egress         ObservationPoint `json:"egress"`
-	Host           string           `json:"host"`
-	TCPFlags       []string         `json:"tcp_flags,omitempty"`
-	NextHop        NextHop          `json:"next_hop,omitempty"`
+	FlushTimestamp   int64            `json:"flush_timestamp"`
+	FlowType         string           `json:"type"`
+	SamplingRate     uint64           `json:"sampling_rate"`
+	Direction        string           `json:"direction"`
+	Start            uint64           `json:"start"` // in seconds
+	End              uint64           `json:"end"`   // in seconds
+	Bytes            uint64           `json:"bytes"`
+	Packets          uint64           `json:"packets"`
+	EtherType        string           `json:"ether_type,omitempty"`
+	IPProtocol       string           `json:"ip_protocol"`
+	Device           Device           `json:"device"`
+	Exporter         Exporter         `json:"exporter"`
+	Source           Endpoint         `json:"source"`
+	Destination      Endpoint         `json:"destination"`
+	Ingress          ObservationPoint `json:"ingress"`
+	Egress           ObservationPoint `json:"egress"`
+	Host             string           `json:"host"`
+	TCPFlags         []string         `json:"tcp_flags,omitempty"`
+	NextHop          NextHop          `json:"next_hop,omitempty"`
+	AdditionalFields AdditionalFields `json:"additional_fields,omitempty"`
+}
+
+// MarshalJSON Custom marshaller that moves AdditionalFields to the root of the payload
+func (p FlowPayload) MarshalJSON() ([]byte, error) {
+	fields := map[string]any{
+		"flush_timestamp": p.FlushTimestamp,
+		"type":            p.FlowType,
+		"sampling_rate":   p.SamplingRate,
+		"direction":       p.Direction,
+		"start":           p.Start,
+		"end":             p.End,
+		"bytes":           p.Bytes,
+		"packets":         p.Packets,
+		"ip_protocol":     p.IPProtocol,
+		"device":          p.Device,
+		"exporter":        p.Exporter,
+		"source":          p.Source,
+		"destination":     p.Destination,
+		"ingress":         p.Ingress,
+		"egress":          p.Egress,
+		"host":            p.Host,
+		"next_hop":        p.NextHop,
+	}
+
+	// omit empty
+	if p.EtherType != "" {
+		fields["ether_type"] = p.EtherType
+	}
+
+	// omit empty
+	if p.TCPFlags != nil {
+		fields["tcp_flags"] = p.TCPFlags
+	}
+
+	// Adding additional fields
+	for k, v := range p.AdditionalFields {
+		if _, ok := fields[k]; ok {
+			// Do not override, override is handled in goflowlib/convert.go
+			continue
+		}
+		fields[k] = v
+	}
+
+	return json.Marshal(fields)
```

Comment:
Won't we lose the additional fields if we ever need to unmarshal one of these in go?

Reply:
This marshaller is used only to get the payload that we then send to DD backend. Backend has no knowledge of this struct and will simply unmarshall all JSON fields.
I don't think there's use case for unmarshalling these back in the agent but let me know if you think of one